objects/69ead2/contained/DUESETUP/script.lua

```diff
@@ -0,0 +1,643 @@
+        direction = Vector(0, 1, 0),
+        max_distance = 1,
+        type = 1, --ray
+        -- debug = true
+    })
+    for _, hit in pairs(hits) do
+        if hit.hit_object.hasTag("Presence") then
+            return 1
+        end
+    end
+    return 2
+end
+
+function onGainPay(params)
+    local dances = Global.call("getSpirit", {name = spiritName})
+    if params.color == Global.call("getSpiritColor", {name = spiritName}) then
+        if params.isGain then
+            for _,tbl in pairs(impendTable) do
+                if params.isUndo then
+                    impendTable[tbl.guid].energy[2] = 0
+                elseif tbl.turnSelected ~= Global.getVar("turn") then
+                    impendTable[tbl.guid].energy[2] = energyPerTurn(dances)
+                end
+                updateEnergyDisplay(tbl.guid)
+            end
+            updateSave()
+        else
+            local zone = Global.getVar("selectedColors")[params.color].zone
+            if params.isUndo then
+                for _,obj in pairs(zone.getObjects()) do
+                    local tbl = impendTable[obj.guid]
+                    if tbl then
+                        -- originally pulled all impend cards back to impend hand but that can be annoying
+                        -- swapping to only pull back cards not currently with sufficient energy
+                        if tbl.energy[1] + tbl.energy[2] + tbl.energy[3] < tbl.maxEnergy then
+                            obj.deal(1, params.color, 3)
+                        end
+                    end
+                end
+            else
+                -- Store all impended cards with enough energy into a table
+                local toPlayFast = {}
+                local toPlaySlow = {}
+                local impendInPlay = {}
+                for _,obj in pairs(zone.getObjects()) do
+                    local tbl = impendTable[obj.guid]
+                    if tbl then
+                        if tbl.energy[1] + tbl.energy[2] + tbl.energy[3] >= tbl.maxEnergy then
+                            impendInPlay[obj] = true
+                        else
+                            obj.deal(1, params.color, 3)
+                        end
+                    end
+                end
+                for _,tbl in pairs(impendTable) do
+                    local card = getObjectFromGUID(tbl.guid)
+                    local energy = tbl.energy[1] + tbl.energy[2] + tbl.energy[3]
+                    if tbl.turnSelected ~= Global.getVar("turn") and energy >= tbl.maxEnergy then
+                        -- Check the card is not in a bag/deck
+                        if card then -- possibly also check card is not in zone as may annoy people if their cards keep moving
+                            local card_z = card.getPosition().z
+                            local zone_z = zone.getPosition().z
+                            if not (impendInPlay[card] and card_z > zone_z and card_z < zone_z + 9.5) then
```

Comment:
i'm not really following the comment or this conditional to understand what case this is trying to avoid

Reply:
When the player pays the energy for the turn, all impended cards that have enough energy go into the play area automatically. The conditional part is to prevent any cards that are already in the play area from being moved into the play area (since they are already there) since the player may have laid out their cards in a certain way. A more specific example is if the player has already pressed pay, then undid their pay, instead of pulling back all of the cards that were put into play through being impended, it leaves the ones that have hit their threshold there to reduce how much back and forth of cards there is. And again when they press play, the cards that are in play are ignored for the movement purely to reduce cards flying all over the place.

Reply:
makes sense, thanks